---
title: "R Notebook"
output: html_notebook
---
# ***setup
```{r}
library(MetBrewer)
library(tidyverse)
library(dada2)
library(phyloseq)
library(conflicted)
library(pheatmap)
library(wesanderson)
library(ggtreeExtra)
library(ggtree)
library(treeio)
library(tidytree)
library(ggstar)
library(ggnewscale)
library(TDbook)
conflicts_prefer(base::setdiff)
conflict_prefer("select", "dplyr")
conflicts_prefer(dplyr::filter)
conflicts_prefer(dplyr::filter)
pcData=function(ps,pcaPlot,nPCs=10) {
  pc.data=ps@sam_data %>%
  data.frame() %>%
  mutate(name=row.names(.)) %>%
  merge(pcaPlot$data[,1:(nPCs+1)],by='name')
}
parseGLMs=function(models) { 
glm.results=data.frame(
    term=character(0),
    estimate=numeric(0),
    std.error=numeric(0),
    statistic=numeric(0),
    p.value=numeric(0),
    formula=character(0),
    message=character(0) 
  )
    for (i in seq_along(models)) { 
      output=models[[i]]
  tidy_output=broom::tidy(output)
  formula=formula(output)
  tidy_output$formula=paste(deparse(formula),collapse='')
  tidy_output
  tidy_output$message=names(models[i])
  glm.results=bind_rows(glm.results,tidy_output)
    }
return(glm.results)
}


pcaPlot<-function(ps, # clr transformed and filtered data
                  colorVar, # variable from samdf to color samples by
                  colorName, # what to display variable name as in legend
                  nTaxa # number of taxa to display
                  ) { 
  if("name" %in% colnames(data.frame(ps@sam_data) )) { # there will be an error if there is a column called "name" in the ps object
  sample_data(ps)=ps@sam_data%>%
    data.frame() %>%
    dplyr::rename(name.x=name)
  }
  samdf <- data.frame(ps@sam_data)%>%
    rownames_to_column(var = 'name')
  
  # PCA
  pca <- prcomp(ps@otu_table, center = TRUE, scale = FALSE)
  
  pca.df <- data.frame(pca$x) %>% 
          rownames_to_column(var = 'name')
  
  # % variance explained
  eigs <- pca$sdev^2
  varExplained <- 100 * round(eigs/sum(eigs), 5)
  
  # pull out first 2 PC variance explained for the plot
  ve.pc1 <- as.character(round(varExplained[1], 3))
  ve.pc2 <- as.character(round(varExplained[2], 3))
  
  # Add back sample data
  pca.df <- left_join(pca.df, samdf)
  
  # Calculate plotting limits based on largest value observed in PC axes 1 and 2
  limit <- max(abs(pca.df[, c('PC1', 'PC2')])) +
            0.05*(max(abs(pca.df[, c('PC1', 'PC2')])))
  
  

  pca.plot <- 
       ggplot(pca.df, aes_string(x = "PC1", y = "PC2", color = colorVar)) +
       geom_point(size = 2, alpha = 0.5) +
       coord_equal() +
       labs(x = paste0(' PC1 (', ve.pc1, '%)'),
            y = paste0(' PC2 (', ve.pc2, '%)')) + 
       xlim(-limit, limit) + ylim(-limit, limit)+
       theme_classic() +
       theme(axis.line = element_line(size = 1, color = 'black'),
             axis.ticks = element_line(color = 'black'),
             axis.title = element_text(size = 14, face = 'bold', color = 'black'),
             # axis.text = element_blank(),
             # legend.background = element_blank(),
             # legend.title = element_blank(),
             #legend.position = "none"
             # legend.text = element_text(size = 10, face = 'bold'),
             ) 

 pca.plot
  # Biplot
  
  # Calculate loadings
  V <- pca$rotation # Eigenvectors
  L <- diag(pca$sdev) # Diag mtx w/sqrts of eigenvalues on diag.
  loadings <- V %*% L
       
  # Get loadings for first 2 PCs and format for plotting
  pythag <- function(a, b){sqrt(a^2 + b^2)}
  loadings.12 <- data.frame(loadings[, 1:2]) %>%
       dplyr::rename(PC1 = X1, PC2 = X2) %>% 
       mutate(variable = row.names(loadings)) %>% 
       mutate(length = pythag(PC1, PC2), slope = PC2/PC1, ang = atan(slope)*(180/pi))
  
  loadings.plot <- top_n(loadings.12, nTaxa, wt = length) 
  
   loadings.plot <- top_n(loadings.12, nTaxa, wt = length) 
  
  # Rename loadings with lowest taxonomic level
  loadings.taxtab <- tax_table(ps)[row.names(loadings.plot)] %>% 
       data.frame() 
  loadings.taxtab <- loadings.taxtab[cbind(1:nrow(loadings.taxtab), max.col(!is.na(loadings.taxtab), ties.method = 'last'))] %>%  
    data.frame()
  colnames(loadings.taxtab) <- c("name")
  loadings.taxtab$asv <- tax_table(ps)[row.names(loadings.plot)] %>% 
    data.frame() %>% 
    rownames()
  
  
  loadings.plot <- loadings.taxtab %>% 
       dplyr::select(asv,name) %>% 
       right_join(loadings.plot, by = c('asv' = 'variable'))
  
  # What quadrant of the plot is the label in?
  q1 <- filter(loadings.plot, PC1 > 0 & PC2 > 0)
  q2 <- filter(loadings.plot, PC1 < 0 & PC2 > 0)
  q3 <- filter(loadings.plot, PC1 < 0 & PC2 < 0)
  q4 <- filter(loadings.plot, PC1 > 0 & PC2 < 0)
       
  pca.biplot <- 
       pca.plot + 
       geom_segment(data = loadings.plot,
                    aes(x = 0, y = 0, 
                        xend = PC1, yend = PC2),
                    color = 'black',
                    arrow = arrow(angle = 15, 
                                  length = unit(0.1, 'inches'))) + 
    labs(color = colorName)
  
  # Then add geom_text quadrant-by-quadrant, aligning text accordingly
       if (dim(q1)[1] != 0) {
            pca.biplot <- pca.biplot +
                 geom_text(data = q1, aes(x = PC1, y = PC2, hjust = 0, angle = ang,
                                          label=paste0('   ', name),
                                          fontface = 'bold'),
                           color = 'black', show.legend = FALSE)
       }
       if (dim(q2)[1] != 0) {
            pca.biplot <- pca.biplot +
                 geom_text(data = q2, aes(x = PC1, y = PC2, hjust = 1, angle = ang,
                                          label=paste0(name, '   '),
                                          fontface = 'bold'),
                           color = 'black', show.legend = FALSE)
       }
       if (dim(q3)[1] != 0) {
            pca.biplot <- pca.biplot +
                 geom_text(data = q3, aes(x = PC1, y = PC2, hjust = 1, angle = ang,
                                          label=paste0(name, '   '),
                                          fontface = 'bold'),
                           color = 'black', show.legend = FALSE)
       }
       if (dim(q4)[1] != 0) {
            pca.biplot <- pca.biplot +
                 geom_text(data = q4, aes(x = PC1, y = PC2, hjust = 0, angle = ang,
                                          label=paste0('   ', name),
                                          fontface = 'bold'),
                           color = 'black', show.legend = FALSE)
       }
  
  print(pca.biplot)
}

pcaPlot_noLoadings<-function(ps, # clr transformed and filtered data
                  colorVar, # variable from samdf to color samples by
                  colorName # what to display variable name as in legend
                  ) { 
  if("name" %in% colnames(data.frame(ps@sam_data) )) { # there will be an error if there is a column called "name" in the ps object
  sample_data(ps)=ps@sam_data%>%
    data.frame() %>%
    dplyr::rename(name.x=name)
  }
  samdf <- data.frame(ps@sam_data)%>%
    rownames_to_column(var = 'name')
  
  # PCA
  pca <- prcomp(ps@otu_table, center = TRUE, scale = FALSE)
  
  pca.df <- data.frame(pca$x) %>% 
          rownames_to_column(var = 'name')
  
  # % variance explained
  eigs <- pca$sdev^2
  varExplained <- 100 * round(eigs/sum(eigs), 5)
  
  # pull out first 2 PC variance explained for the plot
  ve.pc1 <- as.character(round(varExplained[1], 3))
  ve.pc2 <- as.character(round(varExplained[2], 3))
  
  # Add back sample data
  pca.df <- left_join(pca.df, samdf)
  
  # Calculate plotting limits based on largest value observed in PC axes 1 and 2
  limit <- max(abs(pca.df[, c('PC1', 'PC2')])) +
            0.05*(max(abs(pca.df[, c('PC1', 'PC2')])))
  
  

  pca.plot <- 
       ggplot(pca.df, aes_string(x = "PC1", y = "PC2", color = colorVar)) +
       geom_point(size = 2, alpha = 0.5) +
       coord_equal() +
       labs(x = paste0(' PC1 (', ve.pc1, '%)'),
            y = paste0(' PC2 (', ve.pc2, '%)')) + 
       xlim(-limit, limit) + ylim(-limit, limit)+
       theme_classic() +
       theme(axis.line = element_line(size = 1, color = 'black'),
             axis.ticks = element_line(color = 'black'),
             axis.title = element_text(size = 14, face = 'bold', color = 'black'),
             # axis.text = element_blank(),
             # legend.background = element_blank(),
             # legend.title = element_blank(),
             #legend.position = "none"
             # legend.text = element_text(size = 10, face = 'bold'),
             ) 

 return(pca.plot)
}

pcaLoadingsPlot<-function(ps, # clr transformed and filtered data
                  nTaxa, 
                  PCs=c(1,2), #which PC loadings to plot. default is PC1 and PC2
                  title=''
                  ) { 
  if("name" %in% colnames(data.frame(ps@sam_data) )) { # there will be an error if there is a column called "name" in the ps object
  sample_data(ps)=ps@sam_data%>%
    data.frame() %>%
    dplyr::rename(name.x=name)
  }
  samdf <- data.frame(ps@sam_data)%>%
    rownames_to_column(var = 'name')
  
  # PCA
  pca <- prcomp(ps@otu_table, center = TRUE, scale = FALSE)

  # Calculate loadings
  V <- pca$rotation # Eigenvectors
  L <- diag(pca$sdev) # Diag mtx w/sqrts of eigenvalues on diag.
  loadings <- V %*% L
       
  # Get loadings for first 2 PCs and format for plotting
  loadings.plot <- data.frame(loadings[, PCs])
  colnames(loadings.plot)=paste0('PC',PCs)
  loadings.plot$variable = row.names(loadings)
  
  # Rename loadings with lowest taxonomic level
  loadings.taxtab <- tax_table(ps)[row.names(loadings.plot)] %>% 
       data.frame() 
  loadings.taxtab <- loadings.taxtab[cbind(1:nrow(loadings.taxtab), max.col(!is.na(loadings.taxtab), ties.method = 'last'))] %>%  
    data.frame()
  colnames(loadings.taxtab) <- c("name")
  loadings.taxtab$asv <- tax_table(ps)[row.names(loadings.plot)] %>% 
    data.frame() %>% 
    rownames()
  
  
  loadings.plot <- loadings.taxtab %>% 
       dplyr::select(asv,name) %>% 
       right_join(loadings.plot, by = c('asv' = 'variable'))%>%
  pivot_longer(contains("PC"),names_to = 'PC',values_to = 'loading') %>%
  group_by(PC) %>%
  slice_max(abs(loading), n = nTaxa) %>%
  mutate(name=ifelse(is.na(name),
                           asv,
                           name))
  library(tidytext)
plot=loadings.plot %>%
  mutate(name = reorder_within(name, abs(loading), PC))%>%
  ggplot(aes(x=loading,y=name))  +
  geom_bar(stat = "identity", position = "dodge") +
  labs(title = title, x = "Loading Value", y = "") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        legend.position = 'none') + facet_wrap(~PC,scales='free',ncol=1) + 
  scale_y_reordered() 
return(plot)
}
```

## open data




```{r}
# file paths
parent=getwd()

# trnL
ps.trnL=readRDS('FoodSeq-plant.rds')
# 16S
ps.16S=readRDS('Microbiome-16S.rds')

# 12SV5
ps.12sv5=readRDS('FoodSeq-animal.rds')

```

```{r}
ps.16S@sam_data$reads=sample_sums(ps.16S)
ps.all<- ps.16S %>%# NOTE that this data set has SAGE2 samples removed 
  subset_samples(reads>5000)%>%
   filter_taxa(function(x) sum(x > 0) > 1, prune = TRUE) 

taxdf<-ps.all@tax_table %>%
  data.frame() %>%
  mutate(name=ifelse(!is.na(Species),paste(Genus,Species,sep='_'), # setting "name" to the lowest taxonomic level that was assigned 
                     ifelse(!is.na(Genus),Genus, # there is probably a much more elegant way to do this but oh well
                            ifelse(!is.na(Family),Family,
                                   ifelse(!is.na(Order),Order,Phylum)))))
taxdf$name<-make.unique(taxdf$name,sep='_')
taxdf$name=gsub("[^[:alnum:]]", "_", taxdf$name) # remove special characters
taxdf$name=gsub(" ", "_", taxdf$name) # remove whitespace
taxdf$name=gsub("^X","",taxdf$name)
taxdf$name=gsub("^_","",taxdf$name)


tax_table(ps.all)<-as.matrix(taxdf)
taxa_names(ps.all)=ps.all@tax_table[,8]
```


# prep data

```{r}

ps_filt <- ps.all %>%  # Remove samples that do not have any  reads 
  # remove  samples that had <5000 reads (only 3 in dataset but unreliable 16S data)
  subset_samples(reads>=5000 & 
    age_months <= 36 
                 &diarrhea%in%c('No',NA) # remove kids known to have diarrhea
                 &(is.na(WHZ) | WHZ>=-2) # remove acutely malnourished kids (known to have immature microbiome)
                 )%>% 
  # remove rare taxa present in at least 2 samples
  filter_taxa(function(x) sum(x > 0) > 1, prune = TRUE) %>% 
  prune_samples(sample_sums(.) > 0, .) 

ps_filt <- ps_filt %>% 
  subset_samples(!is.na(age_months) & !is.na(subj)) #need to remove all samples that have missing data
ps_filt

```


# set analysis functions

```{r}

getClusters=function(ps_filt,
                           NumClusters=10, # default 10
                           ra_cutoff=0.005, # default relative abundance cutoff is 0.5%
                           filePrefix='' 
) {
  ps_filt.ra <- transform_sample_counts(ps_filt, function(x){x/sum(x)})
  otu.df=ps_filt.ra@otu_table%>%
  data.frame()%>%
  rownames_to_column()%>%
  pivot_longer(-rowname,names_to='name',values_to='ra')%>%
  merge(data.frame(ps_filt.ra@sam_data)%>%select(age_months,country),by.x='rowname',by.y=0,all.x=TRUE)
otu.df

otu.df.by.age=otu.df%>%
  mutate(age_bin=case_when(age_months<=2 ~ 1,
                           between(age_months,2,4) ~ 3,
                           between(age_months,5,7) ~ 6,
                           between(age_months,8,10)~9,
                           between(age_months,11,13)~12,
                           between(age_months,14,16)~15,
                           between(age_months,17,19)~18,
                           between(age_months,20,22)~21,
                           between(age_months,23,25)~24,
                           age_months>25 ~ 36
                           ))%>%
  group_by(age_bin,name)%>%
  summarise(avg_ra=mean(ra))
otu.df.by.age

plotme=otu.df.by.age%>%
  pivot_wider(names_from='age_bin',values_from = 'avg_ra')%>%
  column_to_rownames(var='name')
plotme$country=NULL
plotme=plotme[ , order(as.numeric(names((plotme))))]


# Step 1: Calculate the variance for each row
row_variances <- apply(plotme, 1, var)

# Step 2: Order the dataframe by decreasing row variances
plotme <- plotme[order(-row_variances), ]
plotme

rowmax=apply(plotme, 1, max)
nonZeroIndices <- which(rowmax >= ra_cutoff) # there is at least 1 age bin with x prevalence of taxon
data_matrix=plotme[nonZeroIndices, ]
row_z_scores <- t(apply(data_matrix, 1, function(x) (x - mean(x)) / sd(x)))

# Convert back to a data frame or matrix with the original column names
row_z_scores_df <- as.data.frame(row_z_scores)
colnames(row_z_scores_df) <- colnames(data_matrix)
row_z_scores_df

dist_matrix <- dist(row_z_scores_df) # t() transposes the matrix for row-wise clustering
hc <- hclust(dist_matrix, method = "average")
plot(hc)
# Step 2: Cut the dendrogram to form clusters

clusters <- data.frame(cutree(hc, k = NumClusters))
colnames(clusters)=c('Cluster')

df_cluster=clusters%>%
  rownames_to_column(var='Feature')%>%
  mutate(Feature=gsub("^X","",Feature))%>%
  mutate(Feature=gsub("^_","",Feature))

df_cluster

plot=pheatmap(plotme[nonZeroIndices, ], 
         cutree_rows = NumClusters, 
         show_colnames = TRUE, 
         cluster_cols  = FALSE,
         annotation_names_row = FALSE,
         annotation_row=df_cluster%>%column_to_rownames(var='Feature')%>%mutate(Cluster=factor(Cluster)),
         clustering_method='average',
         scale='row',
         color = colorRampPalette(c("#0B3954",
                                    "#A0C1B9",
                                    "#F4E8C1"))(100),
         angle_col = 0)
print(plot)
  return(df_cluster)
}
```

```{r}
ps.for.merging=ps_filt
getClusterAbundances=function(df_cluster, # data frame with column labeled "Feature" with taxa names and "Cluster" with cluster assignment
                              ps.for.merging) {
physeq=ps.for.merging
df_counts=data.frame(ps.for.merging@otu_table)
colnames(df_counts)=gsub("^X","",colnames(df_counts))

cluster_taxa_list <- split(df_cluster$Feature, df_cluster$Cluster)

# Iterate over each cluster and merge taxa 
for (cluster in names(cluster_taxa_list)) {
  taxa_to_merge <- cluster_taxa_list[[cluster]]
    physeq <- merge_taxa(physeq, taxa_to_merge)
    taxa_names(physeq)[taxa_names(physeq)%in%taxa_to_merge]=paste0('cluster',cluster)
}

physeq.clr=microbiome::transform(physeq, 'clr') 
df=data.frame(physeq.clr@otu_table)[,grepl('cluster',colnames(data.frame(physeq.clr@otu_table)))]
df
return(df)
}
```




# S3.a) heatmap with ordered colors
```{r}
df_cluster=getClusters(ps_filt,
                       filePrefix = '',
                       ra_cutoff=0.005)
```

```{r}
ps_filt.ra <- transform_sample_counts(ps_filt, function(x){x/sum(x)})
  otu.df=ps_filt.ra@otu_table%>%
  data.frame()%>%
  rownames_to_column()%>%
  pivot_longer(-rowname,names_to='name',values_to='ra')%>%
  merge(data.frame(ps_filt.ra@sam_data)%>%select(age_months,country),by.x='rowname',by.y=0,all.x=TRUE)
otu.df

otu.df.by.age=otu.df%>%
  mutate(age_bin=case_when(age_months<=2 ~ 1,
                           between(age_months,2,4) ~ 3,
                           between(age_months,5,7) ~ 6,
                           between(age_months,8,10)~9,
                           between(age_months,11,13)~12,
                           between(age_months,14,16)~15,
                           between(age_months,17,19)~18,
                           between(age_months,20,22)~21,
                           between(age_months,23,25)~24,
                           age_months>25 ~ 36
                           ))%>%
  group_by(age_bin,name)%>%
  summarise(avg_ra=mean(ra))
otu.df.by.age

plotme=otu.df.by.age%>%
  pivot_wider(names_from='age_bin',values_from = 'avg_ra')%>%
  column_to_rownames(var='name')
plotme$country=NULL
plotme=plotme[ , order(as.numeric(names((plotme))))]


# Step 1: Calculate the variance for each row
row_variances <- apply(plotme, 1, var)

# Step 2: Order the dataframe by decreasing row variances
plotme <- plotme[order(-row_variances), ]
plotme

rowmax=apply(plotme, 1, max)
nonZeroIndices <- which(rowmax >= 0.005) # there is at least 1 age bin with x prevalence of taxon
data_matrix=plotme[nonZeroIndices, ]
row_z_scores <- t(apply(data_matrix, 1, function(x) (x - mean(x)) / sd(x)))

# Convert back to a data frame or matrix with the original column names
row_z_scores_df <- as.data.frame(row_z_scores)
colnames(row_z_scores_df) <- colnames(data_matrix)
row_z_scores_df

dist_matrix <- dist(row_z_scores_df) # t() transposes the matrix for row-wise clustering
hc <- hclust(dist_matrix, method = "average")
plot(hc)
# Step 2: Cut the dendrogram to form clusters
# Specify the number of clusters you want, for example, 10

clusters <- data.frame(cutree(hc, k = 10))
colnames(clusters)=c('Cluster')

df_cluster=clusters%>%
  rownames_to_column(var='Feature')%>%
  mutate(Feature=gsub("^X","",Feature))%>%
  mutate(Feature=gsub("^_","",Feature))

df_cluster

plot=pheatmap(plotme[nonZeroIndices, ], 
         cutree_rows = 10, 
         show_colnames = TRUE, 
         cluster_cols  = FALSE,
         annotation_names_row = FALSE,
         #annotation_colors=mycolors,
         annotation_row=df_cluster%>%column_to_rownames(var='Feature')%>%mutate(Cluster=factor(Cluster)),
         clustering_method='average',
         scale='row',
         color = colorRampPalette(c("#0B3954",
                                    "#A0C1B9",
                                    "#F4E8C1"))(100),
         angle_col = 0)
print(plot)

```

```{r}
cluster.mapping=data.frame(Cluster=c(4,3,6,10,2,8,9,5,1,7),
                           name_ordered=c(seq(1:10)))
cluster.mapping=left_join(df_cluster,cluster.mapping,by='Cluster')%>%select(-Cluster)%>%dplyr::rename(Cluster=name_ordered)


# Assign labels only to Cluster 1, 5, and 9; others remain numbered
map <- cluster.mapping %>%
  left_join(data.frame(Cluster = seq(1:10),
                       colors = met.brewer('VanGogh1',10)), by = 'Cluster') %>%
  mutate(Cluster = case_when(
    Cluster == 1 ~ '1 (Transitional)',   
    Cluster == 5 ~ '5 (Late)',           
    Cluster == 9 ~ '9 (Early)',           
    TRUE ~ as.character(Cluster)  # Other clusters keep their number
  )) %>%
  column_to_rownames(var = 'Feature')

color <- distinct(map %>% rownames_to_column() %>% select(-rowname))
mycolors <- setNames(color$colors, color$Cluster)


map <- map %>%
  select(Cluster) %>%
  mutate(Cluster = factor(Cluster, levels = c(
    '1 (Transitional)', '2', '3', '4', '5 (Late)', '6', '7', '8', '9 (Early)', '10'
  )))

# Update colors to match labeled and numbered clusters
mycolors <- mycolors[c('1 (Transitional)', '2', '3', '4', '5 (Late)', '6', '7', '8', '9 (Early)', '10')]

# Prepare annotation colors for heatmap
annotation_colors <- list(Cluster = mycolors)

# Plot heatmap with updated cluster labels
plot <- pheatmap(plotme[nonZeroIndices, ], 
                 cutree_rows = 10, 
                 show_colnames = TRUE, 
                 cluster_cols  = FALSE,
                 annotation_names_row = FALSE,
                 annotation_colors = annotation_colors,
                 annotation_row = map,
                 clustering_method = 'average',
                 scale = 'row',
                 color = colorRampPalette(c("#0B3954", "#A0C1B9", "#F4E8C1"))(100),
                 angle_col = 0)
plot

ggsave(plot=plot,filename='./heatmap-16s-clusters-annotated.png',
       height=12,
       width=10)
```



```{r}
df_cluster=getClusters(ps_global,
                       filePrefix = '',
                       ra_cutoff=0.005)
```


```{r}
data=getClusterAbundances(df_cluster,
                          ps.all)
data
```

### relabel clusters
```{r}
cluster.mapping=data.frame(Cluster=c(4,3,6,10,2,8,9,5,1,7), # relabel to match heatmap
                           name_ordered=c(seq(1:10)))%>%
  mutate(oldname=paste0('cluster',Cluster),
         newname=paste0('cluster',name_ordered))%>%
  select(oldname,newname)
cluster.mapping
numeric_cols <- data %>% select(where(is.numeric))%>%colnames(.)
numeric_cols
data_renamed=data%>%
  t()%>%
  data.frame()%>%
  rownames_to_column(var='oldname')%>%
  left_join(cluster.mapping,by='oldname')%>%
  mutate(newname=ifelse(is.na(newname),oldname,newname))%>%
  column_to_rownames(var='newname')%>%
  select(-oldname)%>%
  t()%>%
  data.frame()%>%
  mutate(across(all_of(numeric_cols), as.double)) # transform everything back to numeric
data_renamed
```

# S3.b) late microbiome against pfr 
```{r,fig.height=10,fig.width=5}
data=data_all%>%
  mutate(Country=country)%>%
  subset(!is.na(cluster5)&!is.na(pFR))
countries <- c('USA', 'Cambodia', 'Nicaragua','Kenya','Pakistan')

# Initialize an empty data frame to store results
results_df <- data.frame(country = character(),
                         intercept = numeric(),
                         slope_pFR = numeric(),
                         slope_age_months = numeric(),
                         stringsAsFactors = FALSE)

for (coun in countries) {
  if (coun%in%c('USA','Kenya')) {model=glm(cluster5 ~ pFR + age_months, data = filter(data, Country == coun), family = 'gaussian') 
  coefficients <- coef(model)
  slope_age_months = coefficients["age_months"]
    p_value_age_months=summary(model)$coefficients["age_months", 4]
  coeff_bf=NA
  pvalue_bf=NA} 
  if (coun%in%c('Pakistan')){model=glm(cluster5 ~ pFR, data = filter(data, Country == coun), family = 'gaussian') 
  coefficients <- coef(model)
  slope_age_months = NA
    p_value_age_months=NA
  coeff_bf=NA
  pvalue_bf=NA }
  if (coun%in%c('Cambodia','Nicaragua')){
    model <-glm(cluster5 ~pFR + currently_breastfed + age_months + currently_breastfed:age_months + pFR:currently_breastfed, data = filter(data, Country == coun), family = 'gaussian')
    coefficients <- coef(model)
    coeff_bf=coefficients["currently_breastfedYes"]
    pvalue_bf=summary(model)$coefficients["currently_breastfedYes", 4]
    p_value_age_months=summary(model)$coefficients["age_months", 4]
    slope_age_months = coefficients["age_months"]
  }
  print(coun)
  print(summary(model))

    # Append to the results data frame
    results_df <- rbind(results_df, data.frame(country = coun,
                                               intercept = coefficients[1],
                                               slope_pFR = coefficients["pFR"],
                                               
                                               coeff_breastfed= coeff_bf,
                                               p_value_breastfed=pvalue_bf,
                                               p_value_pFR=summary(model)$coefficients["pFR", 4],
                                               slope_age_months = slope_age_months,
                                              p_value_age_months=p_value_age_months))
}

x_values <- seq(0, max(data$pFR), by = 1)

# Create line data with all combinations of country and x
line_data <- crossing(Country = results_df$country, x = x_values) %>%
  left_join(results_df, by = c("Country" = "country")) %>%
  mutate(
    y_pFR = intercept + slope_pFR * x,
    y_age_months = intercept + slope_age_months * x
  ) %>%
  select(Country, x, y_pFR, y_age_months)

# Pivot to long format and label variables
line_data_long <- line_data %>%
  pivot_longer(cols = starts_with("y_"), names_to = "line_type", values_to = "y_value") %>%
  mutate(
    Variable = recode(
      sub("y_", "", line_type),
      "pFR" = "Plant Richness (pFR)",
      "age_months" = "Age (months)"
    )
  )
p=results_df%>%
  pivot_longer(c(p_value_pFR,p_value_age_months,p_value_breastfed),names_to = 'variable',values_to = 'p.value')%>%
  mutate(p.adjust=p.adjust(p.value,method='BH'),
         variable=gsub('p_value_','',variable))%>%
  mutate(sig=case_when(p.adjust >= 0.1 ~ 'NS',
                  p.adjust < 0.1 & p.adjust >= 0.05~ ".",
                  p.adjust < 0.05 & p.adjust >= 0.01~ "*",
                  p.adjust < 0.01 & p.adjust >= 0.001~ "**",
                  p.adjust < 0.001 ~ "***"),
         variable=case_when(variable=='pFR' ~ 'Plant Richness (pFR)',
                            variable=='age_months' ~ 'Age (months)',
                            variable=='breastfed'~'Breastfed'))%>%
  mutate(significance_caption=paste0(variable,": ",sig))%>%
  select(country,variable,p.adjust,significance_caption)%>%
  dplyr::rename(Country=country)

p_values_df <- p %>%
  group_by(Country) %>%
  summarise(significance_caption = paste(significance_caption, collapse = "\n"), .groups = "drop")



data <- data %>%
  mutate(
    age_percentile = if_else(age_months > 24, 30 / 24, age_months / 24)
  )

interp <- approx(data$age_months, data$age_percentile, xout = c(0, 6, 12, 18, 24))

breaks <- interp$y
labels <- interp$x
colours <- met.brewer('Hiroshige', length(breaks))
r <- range(data$age_percentile)

data%>%
ggplot(aes(x = pFR, y = cluster5)) +
 geom_point(aes(color = age_percentile)) +  
  facet_wrap(~Country, ncol = 1) +  
  geom_text(data = unique(p_values_df[, c("Country", "significance_caption")]),
            aes(label = significance_caption, x = Inf, y = Inf),
            hjust = 1.1, vjust = 1.1, check_overlap = TRUE, size = 4,fontface='bold') +
  theme_bw() +
  geom_line(data = line_data_long%>%subset(Variable=='Plant Richness (pFR)'), # only plot pFR line
            aes(x = x, y = y_value, 
                ),
            size=1) + 
  labs(x = "Plant Richness (pFR)", y = "Late Microbiome (cluster5) Abundance (clr transformed counts)", color = "Age (Months)") +
  theme(strip.background = element_rect(fill = "white"),
        legend.position = 'bottom',
        legend.box='vertical',
        legend.text=element_text(family = "Arial",face='plain',size=13),
        text = element_text(family = "Arial",face='bold',size=20))+
  scale_color_gradientn(limits  = r,
                        labels=labels,
                        breaks=breaks,
    colours = colours[c(1, seq_along(colours), length(colours))],
    values  = c(0, scales::rescale(breaks, from = r), 1))+ylim(-1,20)+xlim(0,45)

ggsave('cluster5-byCountry-no-interaction-term.png',
       height=10,
       width=5)
```

##  within country clusters over time       
```{r}
ps_filt <- ps.all %>%  # Remove samples that do not have any  reads 
  # remove  samples that had <5000 reads (only 3 in dataset but unreliable 16S data)
  subset_samples(reads>=5000 
                 & age_months <= 36 
                 &diarrhea%in%c('No',NA) # remove kids known to have diarrhea
                 &(is.na(WHZ) | WHZ>=-2) # remove acutely malnourished kids (known to have immature microbiome)
                 )%>% 
  prune_samples(sample_sums(.) > 0, .) %>%
  filter_taxa(function(x) sum(x > 0) > 1, prune = TRUE) # remove rare taxa (present in only 1 sample)

```



```{r}

ps.16s.glom=ps_filt
df_counts=data.frame(ps_filt@otu_table)
colnames(df_counts)=gsub("^X","",colnames(df_counts))

cluster_taxa_list <- split(df_cluster$Feature, df_cluster$Cluster)

# Iterate over each cluster and merge taxa 
for (cluster in names(cluster_taxa_list)) {
  taxa_to_merge <- cluster_taxa_list[[cluster]]
    ps.16s.glom <- merge_taxa(ps.16s.glom, taxa_to_merge)
    taxa_names(ps.16s.glom)[taxa_names(ps.16s.glom)%in%taxa_to_merge]=paste0('cluster',cluster)
}

data=ps.16s.glom@otu_table%>%
  data.frame()%>%
  rownames_to_column(var='match')%>%
  pivot_longer(-match)%>%
  left_join(data.frame(ps.16s.glom@sam_data)%>%rownames_to_column(var='match')%>%select(match,country,age_months,reads),by='match')%>%
  mutate(age_bin=case_when(age_months<=2 ~ 1,
                           between(age_months,2,4) ~ 3,
                           between(age_months,5,7) ~ 6,
                           between(age_months,8,10)~9,
                           between(age_months,11,13)~12,
                           between(age_months,14,16)~15,
                           between(age_months,17,19)~18,
                           between(age_months,20,22)~21,
                           between(age_months,23,25)~24,
                           age_months>25 ~ 36
                           ))

taxdf=data.frame(ps.16s.glom@tax_table)%>%select(-name)%>%rownames_to_column(var='name')
taxdf
cluster.mapping=data.frame(Cluster=c(4,3,6,10,2,8,9,5,1,7),
                           label=c(seq(1:10)))%>%
  mutate(name=paste0('cluster',Cluster))


data1=data%>%
  left_join(taxdf,by='name')%>%
  left_join(cluster.mapping,by='name')%>%
  mutate(is.cluster=case_when(Cluster==1 ~ 'Early Microbiome\n(Bifidobacterium-Streptococcus)',
                              Cluster==2 ~ 'Late Microbiome\n(Prevotella-Faecalibacterium)',
                              Cluster==4 ~ 'Transitional Microbiome\n(Blautia-Ruminococcus)',
                            Cluster%in%c(3,5,6,7,8,9,10,NA) ~ ''))%>%
  mutate(is.cluster=factor(is.cluster,levels=c('','Early Microbiome\n(Bifidobacterium-Streptococcus)','Transitional Microbiome\n(Blautia-Ruminococcus)','Late Microbiome\n(Prevotella-Faecalibacterium)')),
         age_bin=factor(age_bin)
)%>%
  group_by(match,is.cluster,country,age_bin)%>%
  summarise(value=sum(value),
            reads=reads)%>%
  mutate(ra=value/reads)
data1

```

```{r}
data1%>%
  group_by(age_bin,is.cluster,country)%>%
  summarise(avgRA=mean(ra))%>%
  ggplot(aes(x=age_bin,y=avgRA,fill = is.cluster))+
  geom_col(colour=NA)+
  labs(fill='Cluster',x='Age (Months)',y='Relative abundance')+
  scale_fill_manual(
  values = c("white", met.brewer("Derain", n = 4)[c(1,2,4)])
)+
  scale_y_continuous(labels = scales::percent_format(accuracy = 1))+
theme(
      text=element_text(family = "Arial",size=12,face='bold'),
      axis.title = element_text(family = "Arial",size=12,face='bold'),
      axis.text = element_text(family = "Arial",size=10,face='plain'),
    legend.position = 'none',
    panel.grid.major = element_blank(),
panel.grid.minor = element_blank(),
 panel.background = element_rect(fill = "white", color = NA), 
  panel.border = element_rect(color = "grey", fill = NA)     
    )+
  facet_wrap(~country,ncol=1)
ggsave('MbClusters-by-age-barplot_byCountry.png',
       width=3,
       height=5)
```


