---
title: "Fig. 3"
output: html_notebook
---
# ***setup
```{r}
library(MetBrewer)
library(tidyverse)
library(dada2)
library(phyloseq)
library(conflicted)
conflicts_prefer(base::setdiff)
conflict_prefer("select", "dplyr")
conflicts_prefer(dplyr::filter)
conflicts_prefer(dplyr::filter)
# merge PC data into phyloseq metadata
pcData=function(ps,pcaPlot,nPCs=10) {
  pc.data=ps@sam_data %>%
  data.frame() %>%
  mutate(name=row.names(.)) %>%
  merge(pcaPlot$data[,1:(nPCs+1)],by='name')
}

pcaPlot<-function(ps, # clr transformed and filtered data
                  colorVar, # variable from samdf to color samples by
                  colorName, # what to display variable name as in legend
                  nTaxa, # number of taxa to display
                  arrowLength=1, # multiplier for arrow length
                  showLoadings=TRUE 
                  ) { 
  if("name" %in% colnames(data.frame(ps@sam_data) )) { # there will be an error if there is a column called "name" in the ps object
  sample_data(ps)=ps@sam_data%>%
    data.frame() %>%
    dplyr::rename(name.x=name)
  }
  samdf <- data.frame(ps@sam_data)%>%
    rownames_to_column(var = 'name')
  
  # PCA
  pca <- prcomp(ps@otu_table, center = TRUE, scale = FALSE)
  
  pca.df <- data.frame(pca$x) %>% 
          rownames_to_column(var = 'name')
  
  # % variance explained
  eigs <- pca$sdev^2
  varExplained <- 100 * round(eigs/sum(eigs), 5)
  
  # pull out first 2 PC variance explained for the plot
  ve.pc1 <- as.character(round(varExplained[1], 3))
  ve.pc2 <- as.character(round(varExplained[2], 3))
  
  # Add back sample data
  pca.df <- left_join(pca.df, samdf)
  
  # Calculate plotting limits based on largest value observed in PC axes 1 and 2
  limit <- max(abs(pca.df[, c('PC1', 'PC2')])) +
            0.05*(max(abs(pca.df[, c('PC1', 'PC2')])))
  
  

  pca.plot <- 
       ggplot(pca.df, aes_string(x = "PC1", y = "PC2", color = colorVar)) +
       geom_point(size = 2, alpha = 0.5) +
       coord_equal() +
       labs(x = paste0(' PC1 (', ve.pc1, '%)'),
            y = paste0(' PC2 (', ve.pc2, '%)')) + 
       xlim(-limit, limit) + ylim(-limit, limit)+
       theme_classic() +
       theme(axis.line = element_line(size = 1, color = 'black'),
             axis.ticks = element_line(color = 'black'),
             axis.title = element_text(size = 14, face = 'bold', color = 'black'),
             # axis.text = element_blank(),
             # legend.background = element_blank(),
             # legend.title = element_blank(),
             #legend.position = "none"
             # legend.text = element_text(size = 10, face = 'bold'),
             ) 

  # Biplot
  
  # Calculate loadings
  V <- pca$rotation # Eigenvectors
  L <- diag(pca$sdev) # Diag mtx w/sqrts of eigenvalues on diag.
  loadings <- V %*% L
       
  # Get loadings for first 2 PCs and format for plotting
  pythag <- function(a, b){sqrt(a^2 + b^2)}
  loadings.12 <- data.frame(loadings[, 1:2]) %>%
       dplyr::rename(PC1 = X1, PC2 = X2) %>% 
       mutate(variable = row.names(loadings)) %>% 
       mutate(length = pythag(PC1, PC2), slope = PC2/PC1, ang = atan(slope)*(180/pi))
  
  loadings.plot <- top_n(loadings.12, nTaxa, wt = length) 
  
   loadings.plot <- top_n(loadings.12, nTaxa, wt = length) 
  
  # Rename loadings with lowest taxonomic level
  loadings.taxtab <- tax_table(ps)[row.names(loadings.plot)] %>% 
       data.frame() 
  loadings.taxtab <- loadings.taxtab[cbind(1:nrow(loadings.taxtab), max.col(!is.na(loadings.taxtab), ties.method = 'last'))] %>%  
    data.frame()
  colnames(loadings.taxtab) <- c("name")
  loadings.taxtab$asv <- tax_table(ps)[row.names(loadings.plot)] %>% 
    data.frame() %>% 
    rownames()
  
  
  loadings.plot <- loadings.taxtab %>% 
       dplyr::select(asv,name) %>% 
       right_join(loadings.plot, by = c('asv' = 'variable'))
  
  # What quadrant of the plot is the label in?
  q1 <- filter(loadings.plot, PC1 > 0 & PC2 > 0)
  q2 <- filter(loadings.plot, PC1 < 0 & PC2 > 0)
  q3 <- filter(loadings.plot, PC1 < 0 & PC2 < 0)
  q4 <- filter(loadings.plot, PC1 > 0 & PC2 < 0)
       
  pca.biplot <- 
       pca.plot + 
       geom_segment(data = loadings.plot,
                    aes(x = 0, y = 0, 
                        xend = arrowLength*PC1, yend = arrowLength*PC2),
                    color = 'black',
                    arrow = arrow(angle = 15, 
                                  length = unit(0.1, 'inches'))) + 
    labs(color = colorName)
  
  # Then add geom_text quadrant-by-quadrant, aligning text accordingly
       if (dim(q1)[1] != 0) {
            pca.biplot <- pca.biplot +
                 geom_text(data = q1, aes(x = arrowLength*PC1, y = arrowLength*PC2, hjust = 0, angle = ang,
                                          label=paste0('   ', name),
                                          fontface = 'bold'),
                           color = 'black', show.legend = FALSE)
       }
       if (dim(q2)[1] != 0) {
            pca.biplot <- pca.biplot +
                 geom_text(data = q2, aes(x = arrowLength*PC1, y = arrowLength*PC2, hjust = 1, angle = ang,
                                          label=paste0(name, '   '),
                                          fontface = 'bold'),
                           color = 'black', show.legend = FALSE)
       }
       if (dim(q3)[1] != 0) {
            pca.biplot <- pca.biplot +
                 geom_text(data = q3, aes(x = arrowLength*PC1, y = arrowLength*PC2, hjust = 1, angle = ang,
                                          label=paste0(name, '   '),
                                          fontface = 'bold'),
                           color = 'black', show.legend = FALSE)
       }
       if (dim(q4)[1] != 0) {
            pca.biplot <- pca.biplot +
                 geom_text(data = q4, aes(x = arrowLength*PC1, y = arrowLength*PC2, hjust = 0, angle = ang,
                                          label=paste0('   ', name),
                                          fontface = 'bold'),
                           color = 'black', show.legend = FALSE)
       }
  if (showLoadings==FALSE) { return(pca.plot)} else {return(pca.biplot) }
  
}

pcaLoadingsPlot<-function(ps, # clr transformed and filtered data
                  nTaxa, 
                  PCs=c(1,2), #which PC loadings to plot. default is PC1 and PC2
                  title=''
                  ) { 
  if("name" %in% colnames(data.frame(ps@sam_data) )) { # there will be an error if there is a column called "name" in the ps object
  sample_data(ps)=ps@sam_data%>%
    data.frame() %>%
    dplyr::rename(name.x=name)
  }
  samdf <- data.frame(ps@sam_data)%>%
    rownames_to_column(var = 'name')
  
  # PCA
  pca <- prcomp(ps@otu_table, center = TRUE, scale = FALSE)

  # Calculate loadings
  V <- pca$rotation # Eigenvectors
  L <- diag(pca$sdev) # Diag mtx w/sqrts of eigenvalues on diag.
  loadings <- V %*% L
       
  # Get loadings for first 2 PCs and format for plotting
  loadings.plot <- data.frame(loadings[, PCs])
  colnames(loadings.plot)=paste0('PC',PCs)
  loadings.plot$variable = row.names(loadings)
  
  # Rename loadings with lowest taxonomic level
  loadings.taxtab <- tax_table(ps)[row.names(loadings.plot)] %>% 
       data.frame() 
  loadings.taxtab <- loadings.taxtab[cbind(1:nrow(loadings.taxtab), max.col(!is.na(loadings.taxtab), ties.method = 'last'))] %>%  
    data.frame()
  colnames(loadings.taxtab) <- c("name")
  loadings.taxtab$asv <- tax_table(ps)[row.names(loadings.plot)] %>% 
    data.frame() %>% 
    rownames()
  
  
  loadings.plot <- loadings.taxtab %>% 
       dplyr::select(asv,name) %>% 
       right_join(loadings.plot, by = c('asv' = 'variable'))%>%
  pivot_longer(contains("PC"),names_to = 'PC',values_to = 'loading') %>%
  group_by(PC) %>%
  slice_max(abs(loading), n = nTaxa) %>%
  mutate(name=ifelse(is.na(name),
                           asv,
                           name))
  library(tidytext)
plot=loadings.plot %>%
  mutate(name = reorder_within(name, abs(loading), PC))%>%
  ggplot(aes(x=loading,y=name))  +
  geom_bar(stat = "identity", position = "dodge") +
  labs(title = title, x = "Loading Value", y = "") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        legend.position = 'none') + facet_wrap(~PC,scales='free',ncol=1) + 
  scale_y_reordered() 
return(plot)
}
hiroshige=c(met.brewer('Hiroshige')[c(1,4,6)],'#7DA4CA','#31475E')
```

## open data




```{r}
# file paths
parent=getwd()

# trnL
ps.trnL=readRDS('FoodSeq-plant.rds')
# 16S
ps.16S=readRDS('Microbiome-16S.rds')

# 12SV5
ps.12sv5=readRDS('FoodSeq-animal.rds')

```

# ***Figure 3***
# 3a) microbiome diversity vs age

```{r}
d<-data.frame(ps.trnL@sam_data)%>%
  subset(!is.na(age_months)&!is.na(Shannon_diversity))
x <- d$age_months
y <- as.numeric(d$Shannon_diversity)

fit <- nls(y ~ SSlogis(x, Asym, xmid, scal), 
           data = d, 
           start = list(Asym = max(as.numeric(y)), xmid = mean(x), scal = 1))
summary(fit)
# Generate predictions from the logistic growth curve
pred_df <- data.frame(x = seq(min(x), max(x), length.out = 100))
pred_df$y_pred <- predict(fit, newdata = pred_df)
pred_df$type='Plant Richness'
```

```{r}
residuals <- residuals(fit) # Model residuals
fitted_values <- fitted(fit) # Predicted values from the model
params <- coef(fit)

# Extract parameters
Asym <- params["Asym"]
xmid <- params["xmid"]
scal <- params["scal"]

# Define the first derivative of the logistic function
dlogis = function(x) {
  rate = (Asym * exp(-(x - xmid) / scal)) / (scal * (1 + exp(-(x - xmid) / scal))^2)
  return(rate)
}

# For a rough visual approach:
x_vals <- seq(min(d$age_months), max(d$age_months), length.out = 1000)
plot(x_vals, dlogis(x_vals), type='l', main='First Derivative of Logistic Growth')

# For a more analytical approach, find when the rate decreases to e.g., 50% of its maximum
max_rate <- max(dlogis(x_vals))
threshold_rate <- max_rate * 0.5

# Assuming a decreasing function after the maximum, find the first x where this occurs
end_linear_phase <- min(x_vals[dlogis(x_vals) < threshold_rate])

print(paste('the plateau point of Microbiome Shannon diversity with respect to age is at',round(end_linear_phase),'months'))
```

```{r}

vline.df=data.frame(type=c('Shannon'),
                    age_months=c(round(end_linear_phase)))
```

```{r}

d%>%
  dplyr::mutate(Country=country)%>%
ggplot(aes(x = age_months, y = Shannon_diversity)) +
  geom_jitter(alpha=0.5,aes(color=Country)) +  # Scatter plot of original data
  geom_line(data = pred_df, aes(x = x, y = y_pred, group=type), color = "black", size = .8,alpha=0.8) +  # Line plot of logistic growth curve
  scale_color_manual(values=hiroshige)+
  xlab("Age (Months)") +
  ylab("Microbiome Diversity (Shannon)") +
  theme_classic() +
  ggtitle("")+ 
  geom_vline(data = vline.df, aes(xintercept = age_months), linetype = "dashed", color = "red")+
  theme(text=element_text(size=16,face='bold'),
        axis.text=element_text(size=16,face='plain',family='Arial'),
        legend.position = 'none')
ggsave('shannon-vs-age-logistic-curve.png',
       height=5,
       width=5)
```

# 3b) diet/microbiome by country
```{r}

res=data.frame(country=character(),
               plat=numeric())
line_df=data.frame(country=character(),
                   x=numeric(),
                   y=numeric(),
                   type=character())

for (coun in c('Cambodia','Kenya','USA')) { # countries with sampling spanning ages 0-2
d<-data.frame(ps.trnL@sam_data)%>%
  subset(!is.na(age_months)&!is.na(pFR)&!is.na(Shannon_diversity))%>%
  subset(country%in%c(coun))
x <- d$age_months
y <- as.numeric(d$pFR)

fit <- nls(y ~ SSlogis(x, Asym, xmid, scal), 
           data = d, 
           start = list(Asym = max(as.numeric(y)), xmid = mean(x), scal = 1))
summary(fit)
# Generate predictions from the logistic growth curve
pred_df <- data.frame(x = seq(min(x), max(x), length.out = 100))
pred_df$y <- predict(fit, newdata = pred_df)
pred_df$type='Plant Richness'
pred_df$country=coun
# Assume fit is your fitted model from nls()
params <- coef(fit)

# Extract parameters
Asym <- params["Asym"]
xmid <- params["xmid"]
scal <- params["scal"]

x_vals <- seq(min(d$age_months), max(d$age_months), length.out = 1000)
y_vals=dlogis(x_vals)
threshold_rate <- (max(y_vals) +min(y_vals))/2
threshold_rate
# Assuming a decreasing function after the maximum, find the first x where this occurs

logis.df=data.frame(x=x_vals,
                    y=y_vals)

end_linear_phase=logis.df%>%
  subset(y<threshold_rate)%>%
  arrange(-y)%>%
  dplyr::slice(1:2)%>%
  .$x%>%
  max()

logis.df%>%
  ggplot(aes(x=x,y=y))+geom_line()+theme_minimal()+geom_vline(xintercept = end_linear_phase, color='red')+ggtitle(paste(coun,'plant diversity'))
line_df=bind_rows(line_df,pred_df)
res=bind_rows(res,
              data.frame(country=coun,
                         plat=end_linear_phase,
                         type='Plant Richness'))
print(paste('in',coun,'the plateau point of plant diversity with respect to age is at',round(end_linear_phase),'months'))
}
```

```{r}
for (coun in c('Cambodia','Kenya','USA')) {
  d<-data.frame(ps.trnL@sam_data)%>%
  subset(!is.na(age_months)&!is.na(pFR)&!is.na(Shannon_diversity))%>%
  subset(country%in%c(coun))
############## repeat, but for 16S ######################
x <- d$age_months
y <- as.numeric(d$Shannon_diversity)

fit <- nls(y ~ SSlogis(x, Asym, xmid, scal), 
           data = d, 
           start = list(Asym = max(as.numeric(y)), xmid = mean(x), scal = 1))
summary(fit)
# Generate predictions from the logistic growth curve
pred_df <- data.frame(x = seq(min(x), max(x), length.out = 100))
pred_df$y <- predict(fit, newdata = pred_df)
pred_df$type='Microbiome Diversity'
pred_df$country=coun

# Assume fit is your fitted model from nls()
params <- coef(fit)

# Extract parameters
Asym <- params["Asym"]
xmid <- params["xmid"]
scal <- params["scal"]

# Define the first derivative of the logistic function
dlogis <- function(x) {
  rate <- (Asym * exp((x-xmid)/scal)) / (scal * (1 + exp((x-xmid)/scal))^2)
  return(rate)
}

x_vals <- seq(min(d$age_months), max(d$age_months), length.out = 1000)
y_vals=dlogis(x_vals)
threshold_rate <- (max(y_vals) +min(y_vals))/2
threshold_rate
# Assuming a decreasing function after the maximum, find the first x where this occurs
#end_linear_phase <- min(x_vals[dlogis(x_vals) < threshold_rate])

logis.df=data.frame(x=x_vals,
                    y=y_vals)

end_linear_phase=logis.df%>%
  subset(y<threshold_rate)%>%
  arrange(-y)%>%
  dplyr::slice(1:2)%>%
  .$x%>%
  max()

logis.df%>%
  ggplot(aes(x=x,y=y))+geom_line()+theme_minimal()+geom_vline(xintercept = end_linear_phase, color='red')+ggtitle(paste(coun,'microbiome diversity'))
line_df=bind_rows(line_df,pred_df)

print(paste('in',coun,'the plateau point of microbiome diversity with respect to age is at',round(end_linear_phase),'months'))
line_df=bind_rows(line_df,pred_df)
res=bind_rows(res,
              data.frame(country=coun,
                         plat=end_linear_phase,
                         type='Microbiome Diversity'))

}
```

```{r}
res
```

```{r}
# add line breaks for plotting 
line_df=line_df%>%
   mutate(type=ifelse(type=='Plant Richness','Plant\nRichness\n(pFR)','Microbiome\n\u03B1 Diversity\n(Shannon)'))

res=res%>%
   mutate(type=ifelse(type=='Plant Richness','Plant\nRichness\n(pFR)','Microbiome\n\u03B1 Diversity\n(Shannon)'))

```

## plot
```{r}
d<-data.frame(ps.trnL@sam_data)%>%
  subset(!is.na(age_months)&!is.na(Shannon_diversity))
d%>%
  pivot_longer(c(Shannon_diversity,pFR),names_to = 'type',values_to = 'richness')%>%
   mutate(type=ifelse(type=='pFR','Plant\nRichness\n(pFR)','Microbiome\n\u03B1 Diversity\n(Shannon)'))%>%
  mutate(type = factor(type, levels = rev(c('Plant\nRichness\n(pFR)', 'Microbiome\n\u03B1 Diversity\n(Shannon)'))))%>%
  subset(country%in%c('USA','Kenya','Cambodia'))%>%
ggplot(aes(x = age_months, y = richness,color=country)) +
  geom_jitter(alpha=0.5) +  # Scatter plot of original data
  geom_line(data = line_df, aes(x = x, y = y, group=interaction(country,type)), color = "blue", size = .8,alpha=0.8) +  # Line plot of logistic growth curve
  xlab("Age (Months)") +
  ylab("Diversity") +
  theme_bw() +
  ylab("")+
  facet_grid(type~country,scales='free_y',switch='y')+
  scale_color_manual(values=hiroshige[c(1,2,5)])+
  geom_vline(data = res, aes(xintercept = plat), linetype = "dashed", color = "red")+ 
  labs(color='Country')+
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        strip.placement = "outside",              
  axis.text.y.left = element_text(margin = margin(r = 5)),  
  axis.ticks.y.left = element_line(),         
  axis.line.y.left = element_line(),    
        strip.background = element_blank(),      
        text = element_text(family = "Arial",
                            face='bold',
                            size=22),
        axis.text = element_text(face='plain',size=14,family='Arial'),
        legend.position = 'none')
ggsave('shannon-pmr-vs-age-logistic-curve-coloredCountry-nopaknic.png',
       height=5,
       width=7.5)
```


# 3c) 16S PCA by country
```{r}
p=ps.16S%>%
  prune_samples(sample_sums(.) > 0, .) %>% # Remove samples that have no reads
  prune_taxa(taxa_sums(.) > 0, .)%>% # remove taxa with no reads
  microbiome::transform(., 'clr')  # clr transform
```

```{r}

plot=p %>% 
  pcaPlot(.,'country','Country',
          nTaxa=10,
          arrowLength=4 # constant by which to multiply loading vectors to make more visible on plot
          ) +
  scale_color_manual(values=hiroshige)+labs(color='Country')
plot+ 

  theme(
        text = element_text(family = "Arial",
                            size=20,
                            face=
                              'bold'),
        axis.title = element_text(family = "Arial",
                            size=20,
                            face=
                              'bold'),
        legend.position = 'none')+
  # extend plot so loadings labels are not cut off
  coord_cartesian(xlim=c(-90,50),
                     ylim=c(-65,40))
ggsave('3_PCA_16S-byCountry-wloadings.png',
        height=7,
       width=7)
```


# 3d) 16S PCA by age
```{r}
# need to scale age so that legend colors are linearly distributed 
p@sam_data$scaled_age=p@sam_data$age_months/24
p@sam_data$scaled_age[p@sam_data$age_months>24]=30/24 # cap age range, set any samples that fall above 24mo to 24mo+6mo (keep scale consistent)
data=data.frame(p@sam_data)
age_breakpoints <- c(0,6,12,18,24) 
interpolated_percentiles <- approx(x = data$age_months, y = data$scaled_age, xout = age_breakpoints, method = "linear")

# Show results
interpolated_results <- data.frame(
  age = age_breakpoints,
  Interpolated_Percentile = interpolated_percentiles$y
)

interpolated_results
breaks=interpolated_results$Interpolated_Percentile
labels=interpolated_results$age
colours<-met.brewer('Hokusai3',length(breaks))
r<-c(0, 30 / 24) # range of scaled age data
r

plot=p %>% 
  pcaPlot(.,'scaled_age','Age (months)',nTaxa=10,arrowLength=4) 

update_geom_defaults("text", list(size = 5))

plot+
  scale_color_gradientn(
    limits  = r,
    colours = colours,
    breaks  = breaks,
    labels  = labels,
     values  = c(0, scales::rescale(breaks, from = r), 1)
  ) +
  labs(color='Age (months)')+ 

  theme(
        text = element_text(family = "Arial",
                            size=20,
                            face=
                              'bold'),
        axis.title = element_text(family = "Arial",
                            size=20,
                            face=
                              'bold'),
        legend.position = 'none')+coord_cartesian(xlim=c(-90,50),
                     ylim=c(-65,40))

ggsave('3_PCA_16S-byAgeMonths-wloadings.png',
       height=7,
       width=7)


```

